File: 0114. 二叉树展开为链表/dfs.cpp
// 4ms
class Solution {
public:
    vector<int> sequence;

    void flatten(TreeNode* root) {
        dfs(root);
        TreeNode* pre = nullptr;
        for (auto e: sequence) {
            if (root == nullptr) {
                root = new TreeNode();
                pre->right = root;
            } else {
                root->left = nullptr;
            }
            root->val = e;
            pre = root;
            root = root->right;
        }
    }

    void dfs(TreeNode* root) {
        if (root != nullptr) {
            sequence.push_back(root->val);
            dfs(root->left);
            dfs(root->right);
        }
    }
};

File: 0129. 求根到叶子节点数字之和/DFS.swift
// 8ms
class Solution {
    func sumNumbers(_ root: TreeNode?) -> Int {
        var result = 0
        func dfs(_ pre: Int, _ root: TreeNode) {
            let val = pre * 10 + root.val
            if let left = root.left {
                dfs(val, left)
            }
            if let right = root.right {
                dfs(val, right)
            }
            if root.left == nil && root.right == nil {
                result += val
            }
        }
        if let root = root {
            dfs(0, root)
        }
        return result
    }
}